### PR TITLE
migrated to laminas/laminas-diactoros (#125)

### PR DESCRIPTION
**Pull request by driehle**
| Q               | A
| --------------- | ---
| Bug fix?         | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | #125 
| License         | MIT

#### What's in this PR?

This PR replaces zend/zend-diactoros with laminas/laminas-diactoros. Due to laminas/laminas-zendframework-bridge, there is no BC break.

#### Why?

Because zend/zend-diactoros is officially abandoned.

### CODE REVIEW

File: puli.json [hunk 1]
```diff
@@ -22,7 +22,7 @@
         },
         "2438c2d0-0658-441f-8855-ddaf0f87d54d": {
             "_class": "Puli\\Discovery\\Binding\\ClassBinding",
-            "class": "Http\\Message\\MessageFactory\\GuzzleMessageFactory",
+            "class": "Http\\Message\\MessageFactory\\GuzcuzleMessageFactory",
```
Comment (dbu) [inline puli.json:25]: was this change on purpose?

Comment (driehle) [reply puli.json:25]: No, I'll fix it

Comment (dbu) [reply puli.json]: thanks

File: composer.json [hunk 2]
```diff
@@ -30,15 +30,16 @@
         "ergebnis/composer-normalize": "^2.1",
         "guzzlehttp/psr7": "^1.0",
         "henrikbjorn/phpspec-code-coverage": "^1.0",
+        "laminas/laminas-diactoros": "^2.0",
+        "laminas/laminas-zendframework-bridge": "^1.0",
```
Comment (dbu) [inline composer.json:34]: this is in the require-dev section.

consumers will use this library together with the implementation of their choice. i fear that we need to do this differently. we could either in the factory check if the laminas classes exist and otherwise fallback to zend. or we could leave the existing factories alone and create new LaminasDiactoros...Factory classes.

Comment (driehle) [reply composer.json:34]: Well, the end user would need to do `composer require laminas/laminas-diactoros`. With that, the end user will not have to change anything in their code. The only thing that will not work is if only the legacy package zend/zend-diactoros is required, but not laminas/lamians-diactoros. Don't you think it would be fine to mention this in the docs?

Comment (driehle) [reply composer.json:34]: If you prefer end users not to switch the library name, we could to:
```php
use Laminas\Diactoros\Request as LaminasRequest;
use Zend\Diactoros\Request as ZendRequest;

public function createRequest(...) 
{
    // this call will trigger autoload!
    if (class_exists('Laminas\Diactoros\Request')) {
        return new LaminasRequest(...);
    }

    return new ZendRequest(...);
}
```

Comment (dbu) [reply composer.json:34]: the problem is, if we don't do a new major version, users might just `composer upgrade` and ignore the warning about zend being abandoned. composer will install, but they application is suddenly broken.

so i think we we should do your suggestion. does laminas always provide the BC layer of extending the old zend classes, or only if the additional bc package is installed?

**Comment by dbu** [comment]
if i understand correctly, this means that if one installs only laminas, but not the zend-bridge, the factory now returns a different class. i would hope that users do not rely on the concrete implementations when using this factory (otherwise there is no need to use the factory anyways). so to me this seems fine. 

i suggest we do a new minor version, as it still is more than a bugfix. can you please rebase on the master branch and add something to the changelog (including mentioning the bridge?)

File: composer.json (hunk 2 above)
Comment (dbu) [inline composer.json:34]: why install the bridge for the tests? things would work fine without, right?

Comment (driehle) [reply composer.json:34]: Yes, we could drop the bridge here.

**Comment by driehle** [comment]
> if i understand correctly, this means that if one installs only laminas, but not the zend-bridge, the factory now returns a different class. i would hope that users do not rely on the concrete implementations when using this factory (otherwise there is no need to use the factory anyways). so to me this seems fine.

Yes, you need to have the bridge installed. However, the bridge is required as a dependency of laminas-diactoros (see https://packagist.org/packages/laminas/laminas-diactoros) and will, to my understanding, be included there as a dependency until laminas-diactoros releases a major update (i.e. 3.0). So as soon as someone will have laminas-diactoros 2.x installed, they will have the bridge installed as well.

In the [autoloader](https://github.com/laminas/laminas-zendframework-bridge/blob/1.2.x/src/Autoloader.php#L112) of the bridge you can see that they use [class_alias()](https://www.php.net/class_alias) to create the legacy classes (e.g. ZendRequest). Therefore, if you have laminas-diactoros and the bridge installed, you will actually use LaminasRequest, but an $laminasRequest instanceof ZendRequest will still hold true (and vice versa, tough that is unlikely to happen).

> i suggest we do a new minor version, as it still is more than a bugfix. can you please rebase on the master branch and add something to the changelog (including mentioning the bridge?)

Sure, will do.

**Comment by driehle** [comment]
Rebased and changelog updated. I fixed some broken and missing links in the changelog as well. ;-)

File: puli.json [hunk 3]
```diff
@@ -7,7 +7,7 @@
             "class": "Http\\Message\\MessageFactory\\DiactorosMessageFactory",
             "type": "Http\\Message\\MessageFactory",
             "parameters": {
-                "depends": "Zend\\Diactoros\\Request"
+                "depends": "Laminas\\Diactoros\\Request"
```
Comment (dbu) [inline puli.json:10]: not sure how many people use puli, but i think we should only add the laminas things but not replace zend.

Comment (driehle) [reply puli.json:10]: I am not using puli and also not sure about its usefullness in general, but the way in understand this would be that adding two dependencies would mean that both (AND) are required and not only one (OR). If this is the case, then we should leave it with laminas here.

Comment (dbu) [reply puli.json:10]: @Nyholm do you know if we can make puli detect both the old and the new class?

if we can't i opt to configure it with the old class. when the BC layer is there, that should even still work.

Comment (dbu) [reply puli.json:10]: @Nyholm wdyt?

**Comment by dbu** [comment]
can you please fix the cs issue? https://github.com/php-http/message/pull/134/checks?check_run_id=1246548559

if tobias can clarify the puli usage, we can adjust that, if not i suggest we revert the puli change to not introduce unexpected BC breaks.

**Comment by driehle** [comment]
> can you please fix the cs issue? https://github.com/php-http/message/pull/134/checks?check_run_id=1246548559
> 
> if tobias can clarify the puli usage, we can adjust that, if not i suggest we revert the puli change to not introduce unexpected BC breaks.

@dbu: Rebased on current master, fixed CS issue. Still awaiting the reply of @Nyholm regarding puli.

File: puli.json (hunk 3 above)
Comment (Nyholm) [reply puli.json:10]: Im honestly not sure either.. :/

Comment (driehle) [reply puli.json:10]: Ok, then we should probably stick to the old class name...

**Review by dbu — APPROVED** [review]
thank you!
